File: swift/0235-lowest-common-ancestor-of-a-binary-search-tree.swift
class Solution {
     func lowestCommonAncestor(_ root: TreeNode?, _ p: TreeNode?, _ q: TreeNode?) -> TreeNode? {
        guard let root = root, let p = p, let q = q else { return nil }
        if root.val < p.val && root.val < q.val {
            return lowestCommonAncestor(root.right, p, q)
        } else if root.val > p.val && root.val > q.val {
            return lowestCommonAncestor(root.left, p, q)
        } else {
            return root
        }
        return nil
    }
}

File: scala/0235-lowest-common-ancestor-of-a-binary-search-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */

import annotation.tailrec

object Solution {
    @tailrec
    def lowestCommonAncestor(root: TreeNode, p: TreeNode, q: TreeNode): TreeNode = {
        if (p.value < root.value && q.value < root.value) {
            lowestCommonAncestor(root.left, p, q)
        } else if (p.value > root.value && q.value > root.value) {
            lowestCommonAncestor(root.right, p, q)
        } else {
            root
        }
    }
}